Add finding.yaml and matches.json formats to documentation

goneall opened the pull request:
Signed-off-by: Gary O'Neall <gary@sourceauditor.com>

goneall commented:
@swinslow If you get a chance, can you review this and make sure I got the format right?

swinslow commented on docs/config.md:
```diff
@@ -8,6 +8,10 @@ Within this directory, there will be a subdirectory for each monthly scan, named
 
 When scaffold is run from the command line, it will load in `config.json`; process the requested actions; and output a revised `config.json` file with the updated status details. It will also store a backup copy of the prior `config.json` file in the `backup/` subdirectory for that month, in case there are any errors that result in an invalid `config.json` file being outputted.
 
+In addition to the `config.json` file, a `finding-[project].yaml` file is needed to generate the HTML report.  Details on the file format are below.
```
Believe this should be `findings-` plural, not `finding`?

goneall replied:
Good catch - thanks @swinslow - fixed in latest commit

swinslow commented on docs/config.md:
```diff
@@ -8,6 +8,10 @@ Within this directory, there will be a subdirectory for each monthly scan, named
 
+In addition to the `config.json` file, a `finding-[project].yaml` file is needed to generate the HTML report.  Details on the file format are below.
+
+An optional `matches-[project].json` file provideing information on bulk matches for fossology.  See the details below.
```
minor nit typo, "provideing" => "providing"

goneall replied:
fixed in latest commit

swinslow commented on docs/config.md:
```diff
@@ -131,3 +135,27 @@ The JSON file has the following fields:
+* `priority` Priority of the finding as displayed in the report
+* `matches-license` license name that this finding applies to as it appears in the fossology
+* `matches-path` optional file path that the finding applies to
+* `text` Text used for the finding in the HTML report
```
one additional option -- `title` is an optional field that, if included, will set the title of the ticket in JIRA (where Jira integration is configured). See the Hyperledger findings file for an example.

swinslow commented on docs/config.md:
```diff
@@ -131,3 +135,27 @@ The JSON file has the following fields:
+* `id` Id unique to each finding
+* `priority` Priority of the finding as displayed in the report
+* `matches-license` license name that this finding applies to as it appears in the fossology
+* `matches-path` optional file path that the finding applies to
```
There is also `matches-subproject` -- I _think_ that that will:
* if included on its own, flag a single instance of this finding for that subproject's report but without tying it to any particular files
* if included together with a `matches-path` or `matches-license` option, require an `AND` of both of them in order to flag an instance of this finding

I'm not 100% certain but I _think_ that's how it behaves...

goneall replied:
I added the matches-subproject and also restructured this section a bit to group all the matches fields together.

swinslow commented on docs/config.md:
```diff
@@ -131,3 +135,27 @@ The JSON file has the following fields:
+The `findings-[project].yaml` file (where `[project]` is replaced with the project name used in the config file) contains information used in the generation of the HTML report.  The file contains a list of `findings` with the following fields:
+* `id` Id unique to each finding
+* `priority` Priority of the finding as displayed in the report
+* `matches-license` license name that this finding applies to as it appears in the fossology
```
Note that `matches-license` is also optional. Some findings could match just on a path without caring what the license is.

Also note that the license name will be as it appears _after_ the translations applied after retrieving the SPDX file -- e.g. how it appears on the web report, which might be different from how it appears in Fossology.

goneall replied:
One question - on the translation for the license name - how is that done?  Is there fields in the config file that translate the license names?

goneall replied:
Restructured this section to (hopefully) make it a bit clearer.

swinslow replied:
For the license name translations, yes, this is tracked in the config.json file.

For a given project, look at the `policies` object, which will define one or more named sub-objects for a policy. Within the policy is a `categories` array, and each category in turn contains a `licenses` array.

Each license object contains a `name` property, which is the name that will appear in the reports; and an `aliases` array of 0+ strings. Each of the strings in an `aliases` array, if there are any, is a value from Fossology's SPDX output that would get "translated" to the `name` value for the reports for that project/policy combination.

Here's an example that might make it easier to understand, this shows one of the objects in a `categories` array:

```
{
    "name": "Attribution",
    "licenses": [
        {
            "name": "BSD (version unspecified)",
            "aliases": [
                "LicenseRef-BSD"
            ]
        },
        {
            "name": "BSD-3-Clause",
            "aliases": []
        }
    }
}, . . .
```

In this example, `BSD` is the license value that the user cleared within Fossology. When Fossology creates the SPDX file, it will use the value `LicenseRef-BSD` as the LicenseConcluded field. And with this configuration, when scaffold ingests that SPDX file, it will translate `LicenseRef-BSD` to `BSD (version unspecified)`, and the latter will be what shows up in reports.

This can be particularly useful for representing complex license expressions. If a file is cleared in Fossology with licenses for e.g. "MIT", "Apache-2.0" and "Dual-license", then in the SPDX file Fossology will correctly represent this as "MIT OR Apache-2.0". But if there's 3 or more entries, then Fossology will output it as e.g. "MIT AND Apache-2.0 AND LicenseRef-Dual-license AND GPL-2.0+". Using the `aliases` to translate that into a different expression can help make it possible to actually convey the real expression in the reports.

goneall replied:
Thanks @swinslow for the detailed description.  Makes sense.

swinslow reviewed (changes requested):
Hi @goneall, here are some initial comments. I haven't gotten through the review but wanted to go ahead and flag what I've noticed so far -- will circle back with more feedback!

swinslow commented on docs/config.md:
```diff
@@ -131,3 +135,31 @@ The JSON file has the following fields:
+
+The `findings` elements contains one or more of the following matches fields.  A match is made if ALL of the match conditions are met (e.g. an AND of all matches). 
+* `matches-license` license name that this finding applies to as it appears in the fossology.  Note that the license name will be as it appears after the translations applied after retrieving the SPDX file -- e.g. how it appears on the web report, which might be different from how it appears in Fossology.
+* `matches-path` file path that the finding applies to
```
Sorry, I should have noted a few details here also for `matches-path`:
* it will match if the file path in the findings file is a string match to any substring in a file's full path.
* I _think_ that you can put `!` at the front of a `matches-path` entry to make it so that a file will only match if it does _not_ contain that string in its path. If you do this, note that you'll have to surround the string in quotation marks so that it's valid YAML. findings-lfenergy.yaml should have an example of this.
* You can put `$` at the end of the `matches-path` entry which means that it will only match if the string is at the end of the file's path

https://github.com/lfscanning/scaffold/blob/master/findings.py#L67 and subsequent lines contain the parsing for these options.

goneall replied:
Resolved with the most recent commit

swinslow commented:
@goneall I looked at the remainder and added a few more comments -- hope this is helpful!

goneall commented:
> I looked at the remainder and added a few more comments -- hope this is helpful!

Thanks @swinslow ! - very helpful.

I've updated the docs a bit more and will merge the PR.

swinslow reviewed (commented):
Hmm -- it doesn't look like this comment got made visible for some reason? @goneall I'm re-submitting it in case you didn't see it, just as an FYI. Sorry for the confusion!

goneall commented on docs/config.md:
```diff
@@ -131,3 +135,31 @@ The JSON file has the following fields:
+* `actions`: List of actions with the following fields:
+	* `action`: Action to be taken - one of:
+		* `add`: Add a new element
+		* FIXME: any others?
```
Thanks @swinslow  - I did not see the previous comment.

I added PR #43 to add this to the docs.
